File: bindings/csharp/Kevy.Client/ClusterClient.cs
// Cluster-aware client (contract §3.15). One connection per shard, CRC16-
// slot routed so single-key commands hit their owner shard directly (no
// server -MOVED / forwarding hop). Requires the server in cluster mode.
// Topology discovered once at connect via CLUSTER SLOTS (16384 slots). CRC16
// matches Redis's key_hash_slot exactly so routing agrees with the server.

namespace Kevy;

/// <summary>A CRC16-routed multi-shard client (§3.15).</summary>
public sealed class ClusterClient : IDisposable
{
    private const int NumSlots = 16384;
    private RespConnection[] _shards;
    private readonly ushort[] _slotToShard;

    private ClusterClient(RespConnection[] shards, ushort[] table) { _shards = shards; _slotToShard = table; }

    /// <summary>Connect via a seed node, discover the topology, and open one
    /// connection per shard.</summary>
    public static ClusterClient Connect(string host, ushort port)
    {
        using var seed = RespConnection.Dial(host, port);
        var reply = seed.Request(["CLUSTER"u8.ToArray(), "SLOTS"u8.ToArray()]);
        var ranges = ParseSlots(reply);
        var (nodes, table) = BuildTopology(ranges);
        var shards = new List<RespConnection>(nodes.Count);
        try
        {
            foreach (var n in nodes) shards.Add(RespConnection.Dial(n.Host, n.Port));
        }
        catch
        {
            foreach (var s in shards) s.Dispose();
            throw;
        }
        return new ClusterClient(shards.ToArray(), table);
    }

    /// <summary>Number of distinct shard nodes.</summary>
    public int ShardCount => _shards.Length;

    private RespConnection ShardFor(KevyBytes key) => _shards[_slotToShard[Crc16.KeyHashSlot(key.Raw)]];

    /// <summary>Route a single-key command to the key's owner shard.</summary>
    public Reply RequestKeyed(KevyBytes key, params KevyBytes[] argv) => ShardFor(key).Request(argv.Raw());
    /// <summary>Async <see cref="RequestKeyed"/>.</summary>
    public Task<Reply> RequestKeyedAsync(KevyBytes key, KevyBytes[] argv, CancellationToken ct = default) =>
        ShardFor(key).RequestAsync(argv.Raw(), ct);

    /// <summary>Send a keyless command to shard 0.</summary>
    public Reply RequestUnkeyed(params KevyBytes[] argv) => _shards[0].Request(argv.Raw());
    /// <summary>Async <see cref="RequestUnkeyed"/>.</summary>
    public Task<Reply> RequestUnkeyedAsync(KevyBytes[] argv, CancellationToken ct = default) =>
        _shards[0].RequestAsync(argv.Raw(), ct);

    /// <summary>PING (any shard; accepts +PONG or +OK).</summary>
    public void Ping()
    {
        var r = RequestUnkeyed("PING");
        if (r.Kind == ReplyKind.Simple && (r.Str() == "PONG" || r.Str() == "OK")) return;
        throw r.Kind == ReplyKind.Error ? Err.ReplyError(r) : Err.Unexpected(r);
    }

    /// <summary>PUBLISH (process-global; any shard).</summary>
    public long Publish(KevyBytes channel, KevyBytes message) => Count(RequestUnkeyed("PUBLISH", channel, message));

    /// <summary>SET (key-routed).</summary>
    public void Set(KevyBytes key, KevyBytes value) => Ok(RequestKeyed(key, "SET", key, value));
    /// <summary>SET … PX (key-routed).</summary>
    public void SetWithTtl(KevyBytes key, KevyBytes value, TimeSpan ttl) =>
        Ok(RequestKeyed(key, "SET", key, value, "PX", Argv.I(Argv.DurationMs(ttl))));

    /// <summary>GET (key-routed); null on miss.</summary>
    public byte[]? Get(KevyBytes key)
    {
        var r = RequestKeyed(key, "GET", key);
        return r.Kind switch
        {
            ReplyKind.Bulk => r.Bytes,
            ReplyKind.Nil or ReplyKind.Null => null,
            ReplyKind.Error => throw Err.ReplyError(r),
            _ => throw Err.Unexpected(r),
        };
    }

    /// <summary>INCR (key-routed).</summary>
    public long Incr(KevyBytes key) => Int(RequestKeyed(key, "INCR", key));
    /// <summary>INCRBY (key-routed).</summary>
    public long IncrBy(KevyBytes key, long delta) => Int(RequestKeyed(key, "INCRBY", key, Argv.I(delta)));
    /// <summary>PEXPIRE (key-routed).</summary>
    public bool Expire(KevyBytes key, TimeSpan ttl) => Bool(RequestKeyed(key, "PEXPIRE", key, Argv.I(Argv.DurationMs(ttl))));
    /// <summary>PERSIST (key-routed).</summary>
    public bool Persist(KevyBytes key) => Bool(RequestKeyed(key, "PERSIST", key));
    /// <summary>PTTL (key-routed).</summary>
    public long TtlMs(KevyBytes key) => Int(RequestKeyed(key, "PTTL", key));

    /// <summary>DEL — routed per key and summed (cross-shard OK).</summary>
    public long Del(params KevyBytes[] keys) => PerKeySum("DEL", keys);
    /// <summary>EXISTS — routed per key and summed.</summary>
    public long Exists(params KevyBytes[] keys) => PerKeySum("EXISTS", keys);

    /// <summary>DBSIZE (whole-cluster; server fans out).</summary>
    public long DbSize() => Count(RequestUnkeyed("DBSIZE"));
    /// <summary>FLUSHALL (whole-cluster; server fans out).</summary>
    public void FlushAll() => Ok(RequestUnkeyed("FLUSHALL"));

    private long PerKeySum(string verb, KevyBytes[] keys)
    {
        long total = 0;
        foreach (var k in keys)
        {
            var n = Int(RequestKeyed(k, verb, k));
            if (n > 0) total += n;
        }
        return total;
    }

    // --- CLUSTER SLOTS parsing + topology ----------------------------------

    private readonly record struct Node(string Host, ushort Port);
    private readonly record struct Range(ushort Start, ushort End, string Host, ushort Port);

    private static (List<Node>, ushort[]) BuildTopology(List<Range> ranges)
    {
        if (ranges.Count == 0) throw new KevyProtocolException("CLUSTER SLOTS returned no ranges");
        var nodes = new List<Node>();
        var table = new ushort[NumSlots];
        foreach (var r in ranges)
        {
            var idx = nodes.FindIndex(n => n.Host == r.Host && n.Port == r.Port);
            if (idx < 0) { nodes.Add(new Node(r.Host, r.Port)); idx = nodes.Count - 1; }
            for (int slot = r.Start; slot <= r.End; slot++) table[slot] = (ushort)idx;
        }
        return (nodes, table);
    }

    private static List<Range> ParseSlots(Reply reply)
    {
        static KevyProtocolException Bad() => new("malformed CLUSTER SLOTS reply");
        if (reply.Kind != ReplyKind.Array) throw Bad();
        var outp = new List<Range>(reply.Items.Count);
        foreach (var row in reply.Items)
        {
            if (row.Kind != ReplyKind.Array || row.Items.Count < 3) throw Bad();
            if (row.Items[0].Kind != ReplyKind.Int || row.Items[1].Kind != ReplyKind.Int) throw Bad();
            var start = row.Items[0].Integer; var end = row.Items[1].Integer;
            var node = row.Items[2];
            if (node.Kind != ReplyKind.Array || node.Items.Count < 2) throw Bad();
            if (node.Items[0].Kind is not (ReplyKind.Bulk or ReplyKind.Simple) || node.Items[1].Kind != ReplyKind.Int) throw Bad();
            var port = node.Items[1].Integer;
            if (!InU16(start) || !InU16(end) || !InU16(port)) throw Bad();
            outp.Add(new Range((ushort)start, (ushort)end, node.Items[0].Str(), (ushort)port));
        }
        return outp;
    }

    private static bool InU16(long n) => n is >= 0 and <= 65535;

    private static long Int(Reply r) => r.Kind switch
    {
        ReplyKind.Int => r.Integer,
        ReplyKind.Error => throw Err.ReplyError(r),
        _ => throw Err.Unexpected(r),
    };

    private static long Count(Reply r)
    {
        var n = Int(r);
        return n < 0 ? throw new KevyProtocolException($"expected non-negative count, got {n}") : n;
    }

    private static bool Bool(Reply r) => r.Kind switch
    {
        ReplyKind.Int => r.Integer == 1,
        ReplyKind.Error => throw Err.ReplyError(r),
        _ => throw Err.Unexpected(r),
    };

    private static void Ok(Reply r)
    {
        if (r.Kind == ReplyKind.Simple && r.Str() == "OK") return;
        throw r.Kind == ReplyKind.Error ? Err.ReplyError(r) : Err.Unexpected(r);
    }

    /// <summary>Close every shard connection.</summary>
    public void Dispose()
    {
        foreach (var s in _shards) s.Dispose();
        _shards = Array.Empty<RespConnection>();
    }
}

File: bindings/java/src/main/java/jp/golia/kevy/ClusterClient.java
// ClusterClient — cluster-aware client, remote-only (client-contract §3.15).
// One connection per shard node, CRC16-slot routed so single-key commands hit
// their owner shard directly (no server -MOVED / forwarding hop). Topology is
// discovered once at connect via CLUSTER SLOTS (16384 slots); CRC16 matches
// Redis's key_hash_slot so routing agrees with the server. Requires the server
// in cluster mode.
package jp.golia.kevy;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.OptionalDouble;

public final class ClusterClient implements AutoCloseable {
    private static final int NUM_SLOTS = 16384;

    private final List<RespConn> shards;
    private final int[] slotToShard;

    private ClusterClient(List<RespConn> shards, int[] slotToShard) {
        this.shards = shards;
        this.slotToShard = slotToShard;
    }

    private record SlotRange(int start, int end, String host, int port) {}
    private record Node(String host, int port) {}

    /** Connect via a seed node, discover topology, open one conn per shard. */
    public static ClusterClient connect(String host, int port) {
        RespConn seed = RespConn.dial(host, port);
        List<SlotRange> ranges;
        try {
            ranges = parseSlots(seed.request(Argv.cmd("CLUSTER").add("SLOTS").list()));
        } finally {
            seed.close();
        }
        List<Node> nodes = new ArrayList<>();
        int[] table = new int[NUM_SLOTS];
        buildTopology(ranges, nodes, table);
        List<RespConn> conns = new ArrayList<>(nodes.size());
        try {
            for (Node n : nodes) conns.add(RespConn.dial(n.host(), n.port()));
        } catch (RuntimeException e) {
            for (RespConn c : conns) c.close();
            throw e;
        }
        return new ClusterClient(conns, table);
    }

    public int shardCount() {
        return shards.size();
    }

    private RespConn shardFor(byte[] key) {
        return shards.get(slotToShard[Crc16.keyHashSlot(key)]);
    }

    public Reply requestKeyed(byte[] key, byte[]... argv) {
        return shardFor(key).request(new ArrayList<>(List.of(argv)));
    }

    public Reply requestUnkeyed(byte[]... argv) {
        return shards.get(0).request(new ArrayList<>(List.of(argv)));
    }

    public void ping() {
        Reply r = requestUnkeyed(Bytes.of("PING"));
        if (r instanceof Reply.Simple s) {
            String v = Bytes.str(s.value());
            if ("PONG".equals(v) || "OK".equals(v)) return;
        }
        throw r.isError() ? Errors.fromReplyText(r.payload()) : Errors.unexpected(r, "+PONG/+OK");
    }

    public long publish(byte[] channel, byte[] message) {
        return Decode.intVal(requestUnkeyed(Bytes.of("PUBLISH"), channel, message));
    }

    // ---- key-routed core ----
    public void set(byte[] key, byte[] value) {
        Decode.ok(requestKeyed(key, Bytes.of("SET"), key, value));
    }

    public void setWithTtl(byte[] key, byte[] value, Duration ttl) {
        Decode.ok(requestKeyed(key, Bytes.of("SET"), key, value, Bytes.of("PX"), Bytes.ofLong(Math.max(1, Bytes.toMs(ttl)))));
    }

    public Optional<byte[]> get(byte[] key) {
        return Optional.ofNullable(Decode.optBulk(requestKeyed(key, Bytes.of("GET"), key)));
    }

    public long incr(byte[] key) {
        return Decode.intVal(requestKeyed(key, Bytes.of("INCR"), key));
    }

    public long incrBy(byte[] key, long delta) {
        return Decode.intVal(requestKeyed(key, Bytes.of("INCRBY"), key, Bytes.ofLong(delta)));
    }

    public boolean expire(byte[] key, Duration ttl) {
        return Decode.bool(requestKeyed(key, Bytes.of("PEXPIRE"), key, Bytes.ofLong(Bytes.toMs(ttl))));
    }

    public boolean persist(byte[] key) {
        return Decode.bool(requestKeyed(key, Bytes.of("PERSIST"), key));
    }

    public long ttlMs(byte[] key) {
        return Decode.intVal(requestKeyed(key, Bytes.of("PTTL"), key));
    }

    // ---- per-key routed + summed ----
    public long del(byte[]... keys) {
        return perKeySum("DEL", keys);
    }

    public long exists(byte[]... keys) {
        return perKeySum("EXISTS", keys);
    }

    private long perKeySum(String verb, byte[][] keys) {
        long total = 0;
        for (byte[] k : keys) {
            long n = Decode.intVal(requestKeyed(k, Bytes.of(verb), k));
            if (n > 0) total += n;
        }
        return total;
    }

    // ---- keyless / whole-cluster ----
    public long dbsize() {
        return Decode.intVal(requestUnkeyed(Bytes.of("DBSIZE")));
    }

    public void flushall() {
        Decode.ok(requestUnkeyed(Bytes.of("FLUSHALL")));
    }

    // ---- a few key-routed collection ops (single-key) ----
    public long hset(byte[] key, byte[]... fieldValueFlat) {
        Argv a = Argv.cmd("HSET").add(key).addAll(fieldValueFlat);
        return Decode.intVal(shardFor(key).request(a.list()));
    }

    public Optional<byte[]> hget(byte[] key, byte[] field) {
        return Optional.ofNullable(Decode.optBulk(requestKeyed(key, Bytes.of("HGET"), key, field)));
    }

    public long lpush(byte[] key, byte[]... values) {
        Argv a = Argv.cmd("LPUSH").add(key).addAll(values);
        return Decode.intVal(shardFor(key).request(a.list()));
    }

    public long sadd(byte[] key, byte[]... members) {
        Argv a = Argv.cmd("SADD").add(key).addAll(members);
        return Decode.intVal(shardFor(key).request(a.list()));
    }

    public long zadd(byte[] key, double score, byte[] member) {
        return Decode.intVal(requestKeyed(key, Bytes.of("ZADD"), key, Bytes.ofDouble(score), member));
    }

    public OptionalDouble zscore(byte[] key, byte[] member) {
        Double d = Decode.optFloat(requestKeyed(key, Bytes.of("ZSCORE"), key, member));
        return d == null ? OptionalDouble.empty() : OptionalDouble.of(d);
    }

    /** Set intersection routed by the FIRST key; all keys must share a slot ({tag}). */
    public List<byte[]> sinter(byte[]... keys) {
        return setCombine("SINTER", keys);
    }

    public List<byte[]> sunion(byte[]... keys) {
        return setCombine("SUNION", keys);
    }

    public List<byte[]> sdiff(byte[]... keys) {
        return setCombine("SDIFF", keys);
    }

    private List<byte[]> setCombine(String verb, byte[][] keys) {
        if (keys.length == 0) throw new InvalidInputException("set combine needs ≥ 1 key");
        Argv a = Argv.cmd(verb).addAll(keys);
        return Decode.bulks(shardFor(keys[0]).request(a.list()));
    }

    @Override
    public void close() {
        for (RespConn c : shards) c.close();
    }

    // ---- CLUSTER SLOTS parsing + topology ----
    private static void buildTopology(List<SlotRange> ranges, List<Node> nodes, int[] table) {
        if (ranges.isEmpty()) throw new ProtocolException("CLUSTER SLOTS returned no ranges");
        for (SlotRange r : ranges) {
            int idx = -1;
            for (int i = 0; i < nodes.size(); i++) {
                Node n = nodes.get(i);
                if (n.host().equals(r.host()) && n.port() == r.port()) { idx = i; break; }
            }
            if (idx < 0) {
                nodes.add(new Node(r.host(), r.port()));
                idx = nodes.size() - 1;
            }
            for (int slot = r.start(); slot <= r.end(); slot++) table[slot] = idx;
        }
    }

    private static List<SlotRange> parseSlots(Reply reply) {
        List<Reply> rows = reply.items();
        if (rows == null) throw new ProtocolException("malformed CLUSTER SLOTS reply");
        List<SlotRange> out = new ArrayList<>(rows.size());
        for (Reply row : rows) {
            List<Reply> cells = row.items();
            if (cells == null || cells.size() < 3) throw new ProtocolException("malformed CLUSTER SLOTS row");
            long start = asInt(cells.get(0));
            long end = asInt(cells.get(1));
            List<Reply> node = cells.get(2).items();
            if (node == null || node.size() < 2) throw new ProtocolException("malformed CLUSTER SLOTS node");
            String host = Bytes.str(node.get(0).payload());
            long port = asInt(node.get(1));
            out.add(new SlotRange((int) start, (int) end, host, (int) port));
        }
        return out;
    }

    private static long asInt(Reply r) {
        if (r instanceof Reply.Int i) return i.value();
        throw new ProtocolException("expected integer in CLUSTER SLOTS, got " + r.shape());
    }
}
